backend/src/route.rs

```diff
@@ -121,13 +139,45 @@ impl Router {
         }
     }
 
-    pub fn route(&self, map: &MapModel, pt1: Coord, pt2: Coord) -> Option<Route> {
-        let start = map.closest_road.nearest_neighbor(&pt1.into()).unwrap().data;
-        let end = map.closest_road.nearest_neighbor(&pt2.into()).unwrap().data;
-        self.route_from_roads(start, end)
+    pub fn route_from_points(
+        &self,
+        router_input: &impl RouterInput,
+        pt1: Coord,
+        pt2: Coord,
+    ) -> Option<Route> {
+        self.route_from_positions(
+            router_input,
+            router_input.snap_to_road(pt1),
+            router_input.snap_to_road(pt2),
+        )
     }
 
-    pub fn route_from_roads(&self, start: RoadID, end: RoadID) -> Option<Route> {
+    /// Routes from the middle of `start` to the middle of `end`.
+    pub fn route_from_roads(
+        &self,
+        router_input: &impl RouterInput,
+        start: RoadID,
+        end: RoadID,
+    ) -> Option<Route> {
+        self.route_from_positions(
+            router_input,
+            Position {
+                road: start,
+                percent_along: 0.5,
+            },
+            Position {
+                road: end,
+                percent_along: 0.5,
+            },
+        )
+    }
+
+    pub fn route_from_positions(
+        &self,
+        _router_input: &impl RouterInput,
```

Comment:
This PR isn't yet adding the behavior of paying attention to the position along the road. But we will need RouterInput to do that -- when `starts` or `ends` is empty, it means there's a modal filter on the start or end road. We need to get the modal filter, to know its position on the road. We also need to look up roads and intersections to figure out which way we can search.